### ISSUE TITLE
Will pdflatex supported in later version

### ISSUE THREAD

**Issue by DXie123** [plk/biblatex#768]
It seems that biblatex can't be compiled with pdflatex, but I have some code need to use pdflatex

**Comment by josephwright** [comment]
There's no (known) issue with pdfTeX: I think we'll need an example here!

**Comment by DXie123** [comment]
```
\documentclass[UTF8]{article}

\usepackage{ctex}
\usepackage[left=20mm,right=20mm,top=25mm, bottom=15mm]{geometry}
\usepackage[colorlinks=true,pdfstartview=FitH,linkcolor=blue,anchorcolor=violet,citecolor=magenta]{hyperref}
\usepackage[backend=biber,bibstyle=gb7714-2015,%nature,%
citestyle=gb7714-2015%,backref=true%
]{biblatex}

\addbibresource[location=local]{example.bib}

\begin{document}

详\cite{王夫之1845--}见文献

\printbibliography

\end{document} 
```

and example.bib looks like this

```
@Book{王夫之1845--,
  Title                    = {宋论},
  Author                   = {王夫之},
  Edition                  = {刻本},
  Publisher                = {曾氏},
  Location                 = {金陵},
  Year                     = {1845(清同治四年)}
}
```

it works well with XeLaTeX -> biber -> XeLaTeX,
but wrong with PDFLaTeX

**Comment by josephwright** [comment]
@DXie123 Well yes, that probably will go wrong: I'd be _very_ surprised if Biber even tries to deal with CJK input and 8-it engines. Really this is a job for a proper Unicode set up (XeTeX or LuaTeX.)

**Comment by moewew** [comment]
The following example works for me
```latex
\documentclass[UTF8]{article}

\usepackage{ctex}
\usepackage[backend=biber,style=numeric]{biblatex}


\usepackage{filecontents}
\begin{filecontents}{\jobname.bib}
@Book{lorem,
  Title                    = {宋论},
  Author                   = {王夫之},
  Edition                  = {刻本},
  Publisher                = {曾氏},
  Location                 = {金陵},
  Year                     = {1845(清同治四年)}
}
\end{filecontents}

\addbibresource{\jobname.bib}

\begin{document}
详\cite{lorem}
见文献

\printbibliography
\end{document} 
```

Note that I have replaced the `王夫之1845--` entrykey with a US-ASCII-only key. If you don't use a Unicode engine using non-ASCII chars in entry keys and most internal macros is a huge gamble and not unlikely to go wrong.

I can produce some errors with `style=authoryear` that seem to stem from the `清同治四年` in the `year` field. Again `biblatex` on 8-bit engines is not really equipped to deal with non ASCII-chars in all situations. In this case the handling of the `edition` and `year`/`labelyear` fields are problematic.

I do also get errors with `style=gb7714-2015`, I'm not familiar with that style and I assume that it uses some techniques similar to these of `style=authoryear` that are simply not compatible with non-ASCII chars on 8-bit engines.

The bottom line is: If you need *full* UTF-8, you must use a Unicode engine like XeLaTeX and LuaLaTeX. There might be a few cases where `biblatex` could be improved, but I think the overall approach that was taken will always be limited with 8-bit engines.

**Comment by moewew** [comment]
I think I have traced down the particular error from `style=authoryear` to the `\IfInteger`s in

https://github.com/plk/biblatex/blob/34adc2e895ce3c886354378a8ec55913cc5768f6/tex/latex/biblatex/biblatex.sty#L5736-L5756

Apparently, they don't fare particularly well with `1845(清同治四年)` as
```latex
\IfInteger{1845(清同治四年)}{y}{n}
```
quickly demonstrates. I'd have preferred `biblatex`'s `\blx@imc@ifinteger`/`\ifinteger`, which does actually work here (though even that is not guaranteed, see https://github.com/plk/biblatex/issues/559), but that can't be used because the argument to the two macros is usually not expanded far enough so for `\ifinteger` to work properly.

**Comment by moewew** [comment]
`biblatex` 3.12 has been released and is available in TeX live 2018 and MikTeX now.

That versions features a few improvements w.r.t. some tests that could break with non-ASCII chars in pdfLaTeX. As mentioned before there are some fundamental constraints that we can't work around, so not everything can be implemented in pdfTeX to fully support all of Unicode.

I'm closing this report for now since the issues that could be addressed on the `biblatex` side have been addressed. If you find any other incompatibilities that have not been mentioned before, please open a new issue and include a short example document that reproduces the undesired behaviour with the `biblatex` standard styles.
